pkg/pushsync/pushsync.go

```diff
@@ -77,59 +78,153 @@ func (ps *PushSync) Close() error {
 		if errors.Is(err, topology.ErrWantSelf) {
-			// i'm the closest - nothing to do
+
+			// Store the chunk in the local store
+			_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+			if err != nil {
+				return fmt.Errorf("chunk store: %w", err)
+			}
+			ps.metrics.TotalChunksStoredInDB.Inc()
+
+			// Send a receipt immediately once the storage of the chunk is successfull
+			err = ps.sendReceipt(w, chunk.Address())
+			if err != nil {
+				return fmt.Errorf("send receipt: %w", err)
+			}
 			return nil
 		}
 		return err
 	}
-	if err := ps.sendChunkMsg(ctx, peer, chunk); err != nil {
+
+	// This is a special situation in that the other peer thinks thats we are the closest node
+	// and we think that the sending peer
+	if p.Address.Equal(peer) {
+
+		// Store the chunk in the local store
+		_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+		if err != nil {
+			return fmt.Errorf("chunk store: %w", err)
+		}
+		ps.metrics.TotalChunksStoredInDB.Inc()
+
+		// Send a receipt immediately once the storage of the chunk is successfull
+		err = ps.sendReceipt(w, chunk.Address())
+		if err != nil {
+			return fmt.Errorf("send receipt: %w", err)
+		}
+		return nil
+	}
+
+	// Forward chunk to closest peer
+	streamer, err := ps.streamer.NewStream(ctx, peer, nil, protocolName, protocolVersion, streamName)
+	if err != nil {
+		return fmt.Errorf("new stream: %w", err)
+	}
+	defer streamer.Close()
+
+	wc, rc := protobuf.NewWriterAndReader(streamer)
+
+	if err := ps.sendChunkDelivery(wc, chunk); err != nil {
+		return fmt.Errorf("send chunk: %w", err)
+	}
+	receiptRTTTimer := time.Now()
+
+	receipt, err := ps.receiveReceipt(rc)
+	if err != nil {
+		return fmt.Errorf("receive receipt: %w", err)
+	}
+	ps.metrics.ReceiptRTT.Observe(time.Since(receiptRTTTimer).Seconds())
+
+	// Check if the receipt is valid
+	if !chunk.Address().Equal(swarm.NewAddress(receipt.Address)) {
+		ps.metrics.InvalidReceiptReceived.Inc()
+		return errors.New("invalid receipt")
+	}
+
+	// forward the receipt to the received stream
+	err = ps.sendReceipt(w, chunk.Address())
+	if err != nil {
+		return fmt.Errorf("send receipt: %w", err)
+	}
+
+	return nil
+}
+
+func (ps *PushSync) getChunkDelivery(r protobuf.Reader) (ch pb.Delivery, err error) {
+	if err = r.ReadMsg(&ch); err != nil {
+		ps.metrics.ReceivedChunkErrorCounter.Inc()
+		return ch, err
+	}
+	ps.metrics.ChunksSentCounter.Inc()
+	return ch, nil
+}
+
+func (ps *PushSync) sendChunkDelivery(w protobuf.Writer, chunk swarm.Chunk) (err error) {
+	startTimer := time.Now()
+	if err = w.WriteMsgWithTimeout(timeToWaitForReceipt, &pb.Delivery{
+		Address: chunk.Address().Bytes(),
+		Data:    chunk.Data(),
+	}); err != nil {
 		ps.metrics.SendChunkErrorCounter.Inc()
 		return err
 	}
+	ps.metrics.SendChunkTimer.Observe(time.Since(startTimer).Seconds())
+	ps.metrics.ChunksSentCounter.Inc()
+	return nil
+}
+
+func (ps *PushSync) sendReceipt(w protobuf.Writer, addr swarm.Address) (err error) {
+	receipt := &pb.Receipt{
```

Comment:
i would pull the creation of a new receipt out of this method and call it with a `pb.Receipt` instead of a `swarm.Address`, this way you are not forced to generate a new canonical receipt any time you call the function, but rather you could reuse the same method to both send new receipts but also to forward existing receipts that you got from other peers and need to forward up the cascade